### ISSUE TITLE
[TwoFluid] MassConservationProcess in MPI

### ISSUE THREAD

**Issue by RiedlAndreas** [KratosMultiphysics/Kratos#8775]
**Description**
Using the MassConservationProcess in the TwoFluidApplication for MPI did not work for me. It seemed to calculate for single ranks a wrong "water_volume_error". And made a correction where none is needed, starting from the first correction time_step.

**Problem**
I think the problem is in mass_conservation_process.cpp

https://github.com/KratosMultiphysics/Kratos/blob/76c3179036053e7df2153c88ffccc590b5bb8926/applications/FluidDynamicsApplication/custom_processes/mass_conservation_check_process.cpp#L82-L85

The variable _mTheoreticalNegativeVolume_ is calculated only for the particular rank. But in line 133 it is compared with the actual volume from all ranks. 

**Solution?**
For me it worked to change the line 84 as follows, so it considers all ranks as a reference value.
```
this->mTheoreticalNegativeVolume = this->mInitialNegativeVolume
```

I'm not sure if this is the best way to do. Or does it make more sense to perform the correction rank wise ?

**Comment by rubenzorrilla** [comment]
The total theoretical volume must be global. Having said so, from your description I guess that a `SumAll` clause is missing. You're proposal would only work for the case with no inflow/outflow, that is, with constant volume of water.

I take the chance to point out that during the past months we've deeply studied the mass correction. Long story short, the conclusion is that this mechanism can be greatly improved. Hopefully we will merge some changes in this regard soon @uxuech @RiccardoRossi

**Comment by mpentek** [comment]
@rubenzorrilla so would you prefer for now to leave it as is and solve this issue with the new merge (sometime soon)? Or would you first suggest and prefer a fix for the current issue? We decided to post it as an issue and not a PR (with a concrete solution) as the TwoFluid is not strictly speaking the main task of interest for Andreas's project.

**Comment by rubenzorrilla** [comment]
> @rubenzorrilla so would you prefer for now to leave it as is and solve this issue with the new merge (sometime soon)? Or would you first suggest and prefer a fix for the current issue? We decided to post it as an issue and not a PR (with a concrete solution) as the TwoFluid is not strictly speaking the main task of interest for Andreas's project.

I think that the mass issue will be soon ready to be merged. If you're not in a hurry, I'd prefer to push this forward rather than updating "legacy" stuff. Makes sense for you?

@uxuech : It shouldn't be too much to clean up the new mass stuff right?

**Comment by mpentek** [comment]
> > @rubenzorrilla so would you prefer for now to leave it as is and solve this issue with the new merge (sometime soon)? Or would you first suggest and prefer a fix for the current issue? We decided to post it as an issue and not a PR (with a concrete solution) as the TwoFluid is not strictly speaking the main task of interest for Andreas's project.
> 
> I think that the mass issue will be soon ready to be merged. If you're not in a hurry, I'd prefer to push this forward rather than updating "legacy" stuff. Makes sense for you?
> 
> @uxuech : It shouldn't be too much to clean up the new mass stuff right?

It would be nice to be able to test it by mid June, but otherwise I am also not in favour of patching up legacy code.

**Comment by mpentek** [comment]
@rubenzorrilla @uxuech Is this implemented meanwhile? Could we test it?
Perhaps the issue is not actual anymore if a solution is available.

**Comment by rubenzorrilla** [comment]
The element changes are already merged. The operations we do are implemented in MPI, however we didn't try it yet.

**Comment by RiedlAndreas** [comment]
Okay I tested the current version of the massConservationProcess in MPI. Unfortunately, I still have the same problem,

Here the initial positive and negative volume is calculated for all ranks. The theoretical negative volume is only calculated for the current rank.

https://github.com/KratosMultiphysics/Kratos/blob/8bc45e785208f5481e8dae7ca1dd25dd05beaab0/applications/FluidDynamicsApplication/custom_processes/mass_conservation_check_process.cpp#L82-L85


The water error is calculated in line 133. Therefore the TheoreticalNegativeVolume (representing only the negative volume of the current rank) is compared with the current negativeVolume (neg_vol calculated for all ranks).

This produces a large water error if more than one rank is used.

The inflow and outflow do not matter, cause the change is considered for all ranks.

https://github.com/KratosMultiphysics/Kratos/blob/8bc45e785208f5481e8dae7ca1dd25dd05beaab0/applications/FluidDynamicsApplication/custom_processes/mass_conservation_check_process.cpp#L111-L133

The error is calculated for all ranks, so in my opinion the initial theoretical volume has also be considered for all ranks. Or do I have a mistake in thinking ?

**Comment by ddiezrod** [comment]
> Okay I tested the current version of the massConservationProcess in MPI. Unfortunately, I still have the same problem,
> 
> Here the initial positive and negative volume is calculated for all ranks. The theoretical negative volume is only calculated for the current rank.
> 
> https://github.com/KratosMultiphysics/Kratos/blob/8bc45e785208f5481e8dae7ca1dd25dd05beaab0/applications/FluidDynamicsApplication/custom_processes/mass_conservation_check_process.cpp#L82-L85
> 
> The water error is calculated in line 133. Therefore the TheoreticalNegativeVolume (representing only the negative volume of the current rank) is compared with the current negativeVolume (neg_vol calculated for all ranks).
> 
> This produces a large water error if more than one rank is used.
> 
> The inflow and outflow do not matter, cause the change is considered for all ranks.
> 
> https://github.com/KratosMultiphysics/Kratos/blob/8bc45e785208f5481e8dae7ca1dd25dd05beaab0/applications/FluidDynamicsApplication/custom_processes/mass_conservation_check_process.cpp#L111-L133
> 
> The error is calculated for all ranks, so in my opinion the initial theoretical volume has also be considered for all ranks. Or do I have a mistake in thinking ?

I think you are correct. mTheoreticalNegativeVolume is missing a SumAll. Just as a comment I'd move these calls to the communicator to `ComputeVolumesAndInterface`, one should expect that this function returns already the correct value, not the local values.

**Comment by mpentek** [comment]
Closed in #9486. What would be an appropriate test case?
